scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java

```diff
@@ -12,125 +13,477 @@
-             * For deleted columns, stale base table metadata is not a problem, but it would be a problem if OUR metadata is stale
-             * and base table metadata is newer (so we have a column, but the base doesn't) - can this happen?
-             *
-             */
-            TableMetadata baseTableMetadata = baseTableMetadata();
-            // TODO: this sets it only for value columns (and not e.g. for `cdc$deleted_` columns), but maybe it's enough?
-            ColumnMetadata baseColumnMetadata = baseTableMetadata.getColumn(columnName);
-            if (baseColumnMetadata != null) {
-                baseType = translateColumnDataType(baseColumnMetadata.getType());
+            // We have a non-collection column. The type
+            // in base table is exactly the same.
+            return cdcDataType;
+        }
+    }
+
+    /**
+     * A wrapper for the base table schema.
+     * <p>
+     * The purpose of this class is to mandate
+     * that all access to the base table schema
+     * is made through this class. This is important,
+     * as we must deal with potential inconsistencies
+     * with the CDC log schema.
+     * <p>
+     * All methods should return some information
+     * from the base table schema <i>as it was at
+     * the time of the CDC table schema</i> (possibly
+     * "synthesizing" such information).
+     * <p>
+     * <b>
+     * Important: before you modify
+     * or add a method to this class, make sure
+     * to properly prove that it is safe
+     * under potential inconsistencies.
+     * Copy the table used in other methods
+     * and fill it in.
+     * </b>
+     */
+    private static class BaseTableSchema {
+        private final TableMetadata baseTableSchema;
+        private final Set<String> partitionColumns;
+        private final Set<String> clusteringColumns;
+        private final Set<String> regularColumns;
+
+        public BaseTableSchema(TableMetadata baseTableSchema) {
+            this.baseTableSchema = baseTableSchema;
+            this.partitionColumns = baseTableSchema.getPartitionKey()
+                    .stream().map(ColumnMetadata::getName).collect(Collectors.toSet());
+            this.clusteringColumns = baseTableSchema.getClusteringColumns()
+                    .stream().map(ColumnMetadata::getName).collect(Collectors.toSet());
+            this.regularColumns = baseTableSchema.getColumns().stream()
+                    .map(ColumnMetadata::getName)
+                    .filter(c -> !this.partitionColumns.contains(c))
+                    .filter(c -> !this.clusteringColumns.contains(c))
+                    .collect(Collectors.toSet());
+        }
+
+        /**
+         * Returns the column type for a column name in the base table,
+         * as of the time of CDC schema.
+         * <p>
+         * Proof of correctness under potential base table schema
```

Comment:
But what is the hypothesis being proven?

Reply:
That the class handles selected schema inconsistencies correctly, I presume.

Reply:
In the amended commit changed it to "Correctness of this implementation under possible inconsistencies with the base table schema" and "Correctness under possible base table schema inconsistencies" (term inconsistency is defined "near")